### PR TITLE
Fixed pt lif precision

### PR DESCRIPTION
**Pull request by ackurth-nc**
Issue Number: 
#329 

Objective of pull request:
There is a bug in the bit accurate ProcessModel of the LIF Process resulting from inconsistencies in the Numpy types chosen to represent the variables.
I change the data type so that it works.

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [-] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [-] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type

Please check your PR type:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?
When setting ```bias_mant``` and ```bias_exp``` to certain (seemingly allowed) positive values, the ```effective_bias``` becomes negative.

## What is the new behavior?
```effective_bias``` becomes positive.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Supplemental information
Test was added.

### CODE REVIEW

File: tests/lava/proc/lif/test_models.py [hunk 1]
```diff
@@ -386,6 +386,31 @@ def test_bitacc_pm_impulse_dv(self):
         self.assertListEqual(expected_v_timeseries, lif_v)
         self.assertListEqual(expected_float_v, lif_v_float.tolist())
 
+    def test_bitacc_pm_eff_bias(self):
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_models.py:389]: Can we call this something like `test_bitacc_pm_scaling_of_bias` to make it more explicit?

File: tests/lava/proc/lif/test_models.py [hunk 2]
```diff
@@ -386,6 +386,31 @@ def test_bitacc_pm_impulse_dv(self):
...
+        """
+        Tests fixed point LIF ProcessModel's scaling of threshold.
+        """
+        shape = (1,)  # a single neuron
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_models.py:393]: I think the comment may not be needed here.

Comment (mathisrichter) [reply tests/lava/proc/lif/test_models.py:393]: I would also remove the `shape` parameter and put it directly into the definition of the `LIF`, where all the other parameters are. It is not used anywhere else and is not an important detail to highlight for this test.

File: tests/lava/proc/lif/test_models.py [hunk 3]
```diff
@@ -386,6 +386,31 @@ def test_bitacc_pm_impulse_dv(self):
...
+        num_steps = 1
+        # Set up high threshold and high bias current to check for potential
+        # overflow in effective bias
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_models.py:396]: Missing a full stop at the end.

File: tests/lava/proc/lif/test_models.py [hunk 4]
```diff
@@ -386,6 +386,31 @@ def test_bitacc_pm_impulse_dv(self):
...
+                  vth=2 ** 17)
+
+        rcnd = RunSteps(num_steps=1)
+        rcfg = LifRunConfig(select_tag='fixed_pt')
+
+        lif.run(condition=rcnd, run_cfg=rcfg)
+        lif_v = lif.v.get()[0]
+        lif.stop()
+
+        # Check if lif_v is positive
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_models.py:411]: Missing a full stop at the end.

File: tests/lava/proc/lif/test_models.py [hunk 5]
```diff
@@ -386,6 +386,31 @@ def test_bitacc_pm_impulse_dv(self):
...
+        self.assertGreater(lif_v, 0)
```
Comment (mathisrichter) [inline tests/lava/proc/lif/test_models.py:412]: For explicit bias, decay, and threshold values, you should be able to test against the exact voltage value that you expect here.

File: src/lava/proc/lif/models.py [hunk 6]
```diff
@@ -69,7 +69,7 @@ class AbstractPyLifModelFixed(PyLoihiProcessModel):
     v: np.ndarray = LavaPyType(np.ndarray, np.int32, precision=24)
     du: int = LavaPyType(int, np.uint16, precision=12)
     dv: int = LavaPyType(int, np.uint16, precision=12)
-    bias_mant: np.ndarray = LavaPyType(np.ndarray, np.int16, precision=13)
+    bias_mant: np.ndarray = LavaPyType(np.ndarray, np.int32, precision=13)
```
Comment (mathisrichter) [inline src/lava/proc/lif/models.py:72]: This is a bit unintuitive. We only "allow" 13-bit precision but need the 32 bits to do shifts later on. I know that leaving it at 16-bit is not much better...
But maybe the 32-bit variable could be introduced as a local copy of `bias_mant` inside the `scale_bias` method? That makes it more obvious what it is needed for. And I would put a comment on it to highlight that the 32 bits are important at the point.

**Review by mathisrichter — CHANGES_REQUESTED** [review]
Nice! Only minor things to be addressed. I put in a bit more detail than I normally would.

File: tests/lava/proc/lif/test_models.py (hunk 1 above)
Comment (ackurth-nc) [reply tests/lava/proc/lif/test_models.py]: Thanks for the comment, addressed in 8fa7f87.

File: tests/lava/proc/lif/test_models.py (hunk 2 above)
Comment (ackurth-nc) [reply tests/lava/proc/lif/test_models.py]: Addressed in 8fa7f87.

File: tests/lava/proc/lif/test_models.py (hunk 3 above)
Comment (ackurth-nc) [reply tests/lava/proc/lif/test_models.py]: Addressed in 8fa7f87.

File: tests/lava/proc/lif/test_models.py (hunk 4 above)
Comment (ackurth-nc) [reply tests/lava/proc/lif/test_models.py]: Addressed in 8fa7f87.

File: tests/lava/proc/lif/test_models.py (hunk 5 above)
Comment (ackurth-nc) [reply tests/lava/proc/lif/test_models.py]: Addressed in 8fa7f87.

File: src/lava/proc/lif/models.py (hunk 6 above)
Comment (ackurth-nc) [reply src/lava/proc/lif/models.py]: I agree that this is unintuitive.
I followed your suggestion of introducing a local copy with promoted ```dtype```, see 0b2c64b.

**Review by mathisrichter — APPROVED** [review]

**Review by PhilippPlank — APPROVED** [review]

**Pull request merged**